c++/src/kj/compat/http.c++

```diff
@@ -3973,6 +4015,75 @@ public:
               });
             }
 
+            kj::Maybe<CompressionParameters> compressionParameters;
+            KJ_IF_MAYBE(agreedParameters, responseHeaders.get(HttpHeaderId::SEC_WEBSOCKET_EXTENSIONS)) {
+              kj::StringPtr failed = "Server failed WebSocket handshake: "_kj;
+              // If the response included websocket extensions, let's make sure we only received a
+              // single (valid) `permessage-deflate` extension.
+              if (offeredExtensions == nullptr) {
+                // We've received extensions when we did not send any in the first place.
+                auto msg = kj::str(
+                    failed, "added Sec-WebSocket-Extensions when client did not offer any.");
+                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
+                  502, "Bad Gateway", msg, nullptr
+                });
+              }
+              auto offers = splitPartsWithCondition(*agreedParameters, ',');
+              if (offers.size() != 1) {
+                // We only expect a single `permessage-deflate` extension, so if we have more offers,
+                // the server has either included repeats, or non-`permessage-deflate` extensions.
+                auto msg = kj::str(
+                    failed,
+                    "response included more than one Sec-WebSocket-Extensions value when only one was expected.");
+                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
+                  502, "Bad Gateway", msg, nullptr
+                });
+              }
+              kj::StringPtr finalOffer = offers[0];
+              if (!finalOffer.startsWith("permessage-deflate"_kj)) {
+                // Our client only sends `permessage-deflate`, thus the server included an extension
+                // that we did not ask for.
+                auto msg = kj::str(
+                    failed,
+                    "response included a Sec-WebSocket-Extensions value that was not permessage-deflate.");
+                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
+                  502, "Bad Gateway", msg, nullptr
+                });
+              }
+              // Only thing left to do is verify the parameters of our single extension.
+              auto parameters = splitPartsWithCondition(finalOffer, ';');
```

Comment:
As with the server side, I'd like to see the parsing code separated from the negotiation logic. We should parse down to a struct first, then in a separate function decide if the options in that struct are acceptable.

Reply:
See: https://github.com/capnproto/capnproto/pull/1480#discussion_r903127931